java/src/com/google/idea/blaze/java/run/hotswap/BlazeReloadFileAction.java

```diff
@@ -0,0 +1,232 @@
+
+    public static final String ACTION_ID = "Debugger.ReloadFile";
+
+    private static final Logger LOGGER = Logger.getInstance(BlazeReloadFileAction.class);
+    private static final Pattern INNER_CLASS_PATTERN = Pattern.compile("^(?:\\$\\w+)*$");
+
+    private final AnAction delegate;
+
+    public BlazeReloadFileAction(AnAction delegate) {
+        super(
+                delegate.getTemplatePresentation().getTextWithMnemonic(),
+                delegate.getTemplatePresentation().getDescription(),
+                delegate.getTemplatePresentation().getIcon());
+        this.delegate = delegate;
+    }
+
+    @Override
+    public void actionPerformed(@NotNull AnActionEvent anActionEvent) {
+        Project project = anActionEvent.getProject();
+        VirtualFile vf = anActionEvent.getRequiredData(CommonDataKeys.VIRTUAL_FILE);
+        ImmutableList<Label> targets = SourceToTargetMap.getInstance(project)
+                .getTargetsToBuildForSourceFile(new File(vf.getPath()));
+        Future<BlazeBuildOutputs> blazeBuildOutputsFuture =
+                BlazeBuildService.getInstance(project).buildFile(vf.getName(), targets);
+
+        ApplicationManager.getApplication().executeOnPooledThread(() -> {
+            try {
+                BlazeBuildOutputs outputs = blazeBuildOutputsFuture.get();
+                if (outputs.buildResult.status != BuildResult.Status.SUCCESS) {
+                    LOGGER.debug("Build failed during hotswap action, hotswap will not be performed");
+                    return;
+                }
+
+                if (!DebuggerSettings.RUN_HOTSWAP_NEVER.equals(DebuggerSettings.getInstance().RUN_HOTSWAP_AFTER_COMPILE)) {
+                    Path relative = ReadAction.compute(() ->
+                            ProjectFileIndex.getInstance(project)
+                                    .getSourceRootForFile(vf).toNioPath().
+                                    relativize(vf.toNioPath()));
+                    String jarDirectory = relative.getParent().toString() + Paths.DELIM;
+                    File tempOutputDir;
+                    try {
+                        tempOutputDir = Files.createTempDirectory("IjBazelHotswap").toFile();
+                        LOGGER.debug("Temp dir for HotSwap artifacts is " + tempOutputDir.getPath());
+                        tempOutputDir.deleteOnExit();
+                    } catch (IOException e) {
+                        LOGGER.error("Failed creating temp directory for hotswap", e);
+                        return;
+                    }
+
+                    ProgressManager.getInstance().run(new Task.Backgroundable(project, "Preparing to hotswap", true) {
+                        @Override
+                        public void run(@NotNull ProgressIndicator progressIndicator) {
+                            // We want to be aware of entries like darwin_arm64-fastbuild/bin/some-api/lib-api.jar and not interested in darwin_arm64-fastbuild/external-api-0.0.654.jar
```

Comment:
Why not? Also, does it accomodate for bzlmod-related changes to path structure? Did you test it with bzlmod?

Reply:
Because libraries are not sources and parts of our project has `platfrom_stuff/bin/` which corresponds to `bazel-bin` symlink

Reply:
Then updating the comment to reflect the intent, not the effect, will be very much appreciated. But anyway I'm not sure if that's a future-proof way of achieving what you want. + I'd like to bump my question about bzlmod

Reply:
@dkashyn-sfdc @brian-mcnamara bump

Reply:
Sorry for the delay, Dmitry is out and I'm not exactly sure whats needed. I know Dmitry added it to help with debugging, I was unsure about it as its quite verbose.